---
title: "part2_new.Rmd"
author: "Thi Tran"
date: "5/31/2020"
output: pdf_document
---

```{r}
library(rBLAST)
library(Biostrings)
library(seqinr)
source("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R")
```

Download the whole E.coli set. Used makeblastdb and blast function from rBLAST library to make database. And then we got the number of sequences is 4,140 sequences

```{r}
download.file("ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz")

# uncompress the file
R.utils::gunzip("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz", overwrite=TRUE)

# create the blast database
makeblastdb("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa", dbtype="nucl", "-parse_seqids")
```


# Cau 2

```{r}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/sample.fa", destfile ="sample.fa") # download sample.fa file
all_sequences <-  seqinr::read.fasta("sample.fa") # read all sequences from fa file
my_seq <- all_sequences[[13]] # my id is 13, get sequences 13-th
# get sequence length and calculate the GC content
length_seq <- seqinr::getLength(my_seq)
gc_proprotion <- seqinr::GC(my_seq)
cat("Sequence length: ", length_seq, '\n')
cat("GC Proportion: ", gc_proprotion)
```

# Cau 3

```{r}
res <- myblastn_tab(myseq = my_seq, db = "Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa")
hits <- as.character(res$sseqid[1:3])
```

# Cau 4
```{r}
my_seq_new <- Biostrings::readDNAStringSet('sample.fa')[13] # read my sequence under Biostring format
my_seq_str <- toString(my_seq_new) # extract as a simple string
my_seq_char <- seqinr::s2c(my_seq_str) # convert string to a vector of characters
# create a mutated copy with 100 substitutions
my_seq_char_mut <- mutator(myseq=my_seqchar, 20)

# now create a pairwise alignment
my_seq_mut_ <- DNAString(c2s(my_seq_char_mut)) # create DNAString format from mutation sequence
aln <- Biostrings::pairwiseAlignment(my_seq_new, my_seq_mut_)
num_mismatch <- nmismatch(aln)
cat("Number of mismatches between the original and mutated sequence: ", num_mismatch)

```


```{r}
matching <- function(nmut) {
  db_path <- 'Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa'
  mut_seq_char <- mutator(my_seq_char, nmut)
  mut_seq <- seqinr::c2s(mut_seq_char)
  match_table <- (myblastn_tab(mut_seq, db_path))
  num_match <- dim(match_table)[1]
  if (is.null(num_match)) {
    return(0)
  }else{
    return(1)
  }
}

probs <- c()

for (nmut in 1:length_seq) {
  prob <- mean(replicate(100, matching(nmut)))
  probs[nmut] <- prob
}

number_mutation <- seq(1:length_seq)
nmut_prevent_search <- which(probs == 0)[1] # get the first position that matching probability is zero.
cat("the number of mutation that prevent BLAST search from matching: ", nmut_prevent_search)
cat("the proprotion of mutation that prevent BLAST search from matching: ", nmut_prevent_search / length_seq)

plot(number_mutation, probs, type='l')

```


```{r}
propotion_plot <- number_mutation / length_seq
#propotion_plot <- propotion_plot[1:nmut_prevent_search]
#probs_plot <- probs[1:nmut_prevent_search]
plot(propotion_plot, probs, type='l')
```
